aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -734,5 +736,34 @@ Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
   }
 AT_FORALL_SCALAR_TYPES_EXCEPT_HALF(TENSOR)
 #undef TENSOR
+
+Tensor from_file(std::string filename, c10::optional<bool> shared, c10::optional<long> size, const TensorOptions& options) {
+    AT_CHECK(!options.pinned_memory(), "tensors constructed from a file cannot be pinned");
+    size_t my_size = 0;
+    if (size.has_value())
+    {
+      my_size = size.value();
+    }
+    int flags = 0;
+    if (shared.has_value() && shared.value()) {
+      flags = TH_ALLOCATOR_MAPPED_SHARED;
+    }
+    size_t actual_size = -1;
+    auto dtype = options.dtype();
```

Comment:
You're completely ignoring the rest of options. It's probably better to just make this explicitly take only `DataType` and dispense with all the other arguments. Saves you an assert too.

Reply:
Better, but our code generation currently out of `from_file(str filename, bool? shared=False, int? size=None, *, ScalarType? dtype=None) -> Tensor` creates abomination like 
`from_file(std::string filename, bool? shared=False, int64_t? size=c10::nullopt, *, ScalarType? dtype=c10::nullopt, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)`

Im thinking if we can update options parsing to fold any subset of dtype, layout, device, .. whatever.. into TensorOptions, but also check if something unexpected passed (however this is out of scope for this PR)